File: js/src/verify.js
/**
 * Content verification module (R6).
 *
 * Compares captured markdown content against the original web page
 * to verify completeness and accuracy.
 *
 * Checks: title, headings, paragraphs, code blocks, formulas,
 * blockquote formulas, list items, links, and figure images.
 *
 * Based on reference implementation from:
 * https://github.com/link-foundation/meta-theory/blob/main/scripts/verify.mjs
 *
 * @module verify
 */

/**
 * Normalize text for comparison.
 * Removes extra whitespace and normalizes unicode characters,
 * LaTeX delimiters, and common symbol substitutions.
 *
 * @param {string} text - Input text
 * @returns {string} Normalized text
 */
export function normalizeText(text) {
  return text
    .trim()
    .replace(/\s+/g, ' ')
    .replace(/[\u2000-\u200F\u2028-\u202F]/g, ' ')
    .replace(/\u00A0/g, ' ')
    .replace(/[\u2018\u2019]/g, "'")
    .replace(/[\u201C\u201D]/g, '"')
    .replace(/[\u00D7]/g, 'x')
    .replace(/\\times/g, 'x')
    .replace(/[\u2192\u21A6]/g, '->')
    .replace(/\\to/g, '->')
    .replace(/[\u2212]/g, '-')
    .replace(/\$\$/g, '')
    .replace(/\$/g, '')
    .replace(/\\displaystyle\s*/g, '')
    .replace(/\\text\{([^}]*)\}/g, '$1')
    .replace(/\\\\%/g, '%')
    .replace(/\\%/g, '%')
    .replace(/\\subseteq/g, '\u2286')
    .replace(/\\mathbb\{n\}_0/gi, '\u2115\u2080')
    .replace(/\\in/g, '\u2208')
    .replace(/\\emptyset/g, '\u2205')
    .replace(/\^2/g, '\u00B2')
    .replace(/\^n/g, '\u207F')
    .toLowerCase();
}

/**
 * Normalize code for comparison (more lenient than text).
 *
 * @param {string} text - Code text
 * @returns {string} Normalized code
 */
export function normalizeCode(text) {
  return text
    .trim()
    .replace(/\s+/g, ' ')
    .replace(/[\u2000-\u200F\u2028-\u202F]/g, ' ')
    .replace(/\u00A0/g, ' ')
    .replace(/[\u00D7]/g, 'x')
    .replace(/\\times/g, 'x')
    .replace(/\$\$/g, '')
    .replace(/\$/g, '')
    .toLowerCase();
}

/**
 * Verify that markdown contains the expected web page content.
 *
 * @param {Object} webContent - Extracted web page content
 * @param {string} webContent.title - Article title
 * @param {Object[]} webContent.headings - Array of {level, text}
 * @param {string[]} webContent.paragraphs - Paragraph texts
 * @param {string[]} webContent.codeBlocks - Code block texts
 * @param {string[]} webContent.formulas - Formula texts
 * @param {string[]} webContent.blockquoteFormulas - Formulas in blockquotes
 * @param {string[]} webContent.listItems - List item texts
 * @param {Object[]} webContent.links - Array of {text, href}
 * @param {number[]} webContent.figures - Figure numbers
 * @param {string} markdownText - The markdown to verify
 * @param {Object} [options] - Verification options
 * @param {boolean} [options.verbose=false] - Detailed output
 * @param {number} [options.expectedFigures] - Expected figure count
 * @param {boolean} [options.hasLocalImages=false] - Whether images are localized
 * @returns {Object} Verification result
 */
export function verifyMarkdownContent(webContent, markdownText, options = {}) {
  const { verbose = false, expectedFigures, hasLocalImages = false } = options;
  const normalizedMarkdown = normalizeText(markdownText);
  const missing = {
    title: false,
    headings: [],
    paragraphs: [],
    codeBlocks: [],
    formulas: [],
    blockquoteFormulas: [],
    listItems: [],
    images: 0,
  };

  let totalChecks = 0;
  let passedChecks = 0;
  const details = [];

  // Check title
  if (webContent.title) {
    totalChecks++;
    const normalizedTitle = normalizeText(webContent.title);
    if (normalizedMarkdown.includes(normalizedTitle)) {
      passedChecks++;
      if (verbose) {
        details.push({ type: 'title', status: 'pass' });
      }
    } else {
      missing.title = true;
      if (verbose) {
        details.push({
          type: 'title',
          status: 'fail',
          text: webContent.title,
        });
      }
    }
  }

  // Check headings
  for (const heading of webContent.headings || []) {
    totalChecks++;
    const normalized = normalizeText(heading.text);
    if (normalizedMarkdown.includes(normalized)) {
      passedChecks++;
    } else {
      missing.headings.push(heading.text);
      if (verbose) {
        details.push({
          type: 'heading',
          status: 'fail',
          text: heading.text,
        });
      }
    }
  }

  // Check paragraphs (sample first 5 and last 5)
  const paragraphs = webContent.paragraphs || [];
  const paragraphsToCheck = [
    ...paragraphs.slice(0, 5),
    ...paragraphs.slice(-5),
  ];

  for (const paragraph of paragraphsToCheck) {
    totalChecks++;
    const normalized = normalizeText(paragraph);
    const words = normalized.split(' ').filter((w) => w.length > 2);
    const matchingWords = words.filter((word) =>
      normalizedMarkdown.includes(word)
    );
    const matchRate =
      words.length > 0 ? matchingWords.length / words.length : 0;

    const substringMatch =
      normalized.length > 20 &&
      normalizedMarkdown.includes(
        normalized.substring(0, Math.min(50, normalized.length))
      );

    if (matchRate >= 0.6 || substringMatch) {
      passedChecks++;
    } else {
      missing.paragraphs.push(`${paragraph.substring(0, 100)}...`);
    }
  }

  // Check code blocks (fuzzy matching)
  const normalizedMarkdownForCode = normalizeCode(markdownText);
  for (const code of webContent.codeBlocks || []) {
    totalChecks++;
    const normalizedCodeFull = normalizeCode(code);

    const lines = code
      .split('\n')
      .map((l) => l.trim())
      .filter((l) => l.length > 3 && !/^[{}[\](),;]+$/.test(l));

    const matchingLines = lines.filter((line) => {
      const normalizedLine = normalizeCode(line);
      return normalizedMarkdownForCode.includes(normalizedLine);
    });

    const matchRate =
      lines.length > 0 ? matchingLines.length / lines.length : 1;

    if (
      matchRate >= 0.6 ||
      normalizedMarkdownForCode.includes(normalizedCodeFull)
    ) {
      passedChecks++;
    } else {
      missing.codeBlocks.push(`${code.substring(0, 100)}...`);
    }
  }

  // Check list items (sample first 10)
  const listItemsToCheck = (webContent.listItems || []).slice(0, 10);
  for (const item of listItemsToCheck) {
    totalChecks++;
    const normalized = normalizeText(item);
    const words = normalized.split(' ').filter((w) => w.length > 2);
    const matchingWords = words.filter((word) =>
      normalizedMarkdown.includes(word)
    );
    const matchRate =
      words.length > 0 ? matchingWords.length / words.length : 0;

    const substringMatch =
      normalized.length > 15 &&
      normalizedMarkdown.includes(
        normalized.substring(0, Math.min(40, normalized.length))
      );

    if (matchRate >= 0.6 || substringMatch) {
      passedChecks++;
    } else {
      missing.listItems.push(`${item.substring(0, 100)}...`);
    }
  }

  // Check blockquote formulas
  if (
    webContent.blockquoteFormulas &&
    webContent.blockquoteFormulas.length > 0
  ) {
    for (const formula of webContent.blockquoteFormulas) {
      totalChecks++;
      const normalizedFormula = formula.replace(/\s+/g, ' ').trim();

      const keyParts = normalizedFormula
        .replace(/\\mathbf\{([^}]*)\}/g, '$1')
        .replace(/\\textbf\{([^}]*)\}/g, '$1')
        .replace(/[{}\\]/g, '')
        .split(/\s+/)
        .filter((part) => part.length > 1);

      const blockquoteLines = markdownText.match(/^>.*$/gm) || [];
      let foundInBlockquote = false;

      for (const line of blockquoteLines) {
        if (line.includes('$')) {
          const matchingParts = keyParts.filter((part) =>
            line.toLowerCase().includes(part.toLowerCase())
          );
          if (
            keyParts.length > 0 &&
            matchingParts.length >= Math.min(2, keyParts.length)
          ) {
            foundInBlockquote = true;
            break;
          }
          if (
            line.includes(normalizedFormula) ||
            line.includes(formula) ||
            (formula.length < 20 && line.includes(formula.replace(/\s/g, '')))
          ) {
            foundInBlockquote = true;
            break;
          }
        }
      }

      if (foundInBlockquote) {
        passedChecks++;
      } else {
        missing.blockquoteFormulas.push(formula.substring(0, 100));
      }
    }
  }

  // Check for figure images
  if (hasLocalImages && expectedFigures) {
    const figurePattern =
      /!\[(?:\*\*)?(?:Figure|Рис\.?|Рисунок)\s*\d+[\s\S]*?\]\(images\/figure-\d+\.(png|jpg)\)/gi;
    const figureMatches = markdownText.match(figurePattern) || [];

    totalChecks++;
    if (figureMatches.length >= expectedFigures) {
      passedChecks++;
    } else {
      missing.images = expectedFigures - figureMatches.length;
    }
  }

  // Calculate results
  const passRate = totalChecks > 0 ? passedChecks / totalChecks : 0;
  const hasMissingContent =
    missing.title ||
    missing.images > 0 ||
    Object.values(missing).some((arr) => Array.isArray(arr) && arr.length > 0);

  return {
    totalChecks,
    passedChecks,
    passRate,
    hasMissingContent,
    missing,
    success: !hasMissingContent || passRate >= 0.85,
    details: verbose ? details : undefined,
  };
}

File: rust/src/verify.rs
//! Content verification module (R6).
//!
//! Compares captured markdown content against the original web page
//! to verify completeness and accuracy.
//!
//! Checks: title, headings, paragraphs, code blocks, formulas,
//! blockquote formulas, list items, links, and figure images.
//!
//! Based on reference implementation from:
//! <https://github.com/link-foundation/meta-theory/blob/main/scripts/verify.mjs>

use regex::Regex;
use serde::{Deserialize, Serialize};

/// Heading with level and text.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Heading {
    pub level: u8,
    pub text: String,
}

/// Content extracted from a web page for verification.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct WebContent {
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub headings: Vec<Heading>,
    #[serde(default)]
    pub paragraphs: Vec<String>,
    #[serde(default)]
    pub code_blocks: Vec<String>,
    #[serde(default)]
    pub formulas: Vec<String>,
    #[serde(default)]
    pub blockquote_formulas: Vec<String>,
    #[serde(default)]
    pub list_items: Vec<String>,
    #[serde(default)]
    pub figures: Vec<u32>,
}

/// Missing content detected during verification.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct MissingContent {
    pub title: bool,
    pub headings: Vec<String>,
    pub paragraphs: Vec<String>,
    pub code_blocks: Vec<String>,
    pub formulas: Vec<String>,
    pub blockquote_formulas: Vec<String>,
    pub list_items: Vec<String>,
    pub images: u32,
}

/// Verification options.
#[derive(Debug, Clone)]
pub struct VerifyOptions {
    pub verbose: bool,
    pub expected_figures: Option<u32>,
    pub has_local_images: bool,
}

impl Default for VerifyOptions {
    fn default() -> Self {
        Self {
            verbose: false,
            expected_figures: None,
            has_local_images: false,
        }
    }
}

/// Result of content verification.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VerifyResult {
    pub total_checks: u32,
    pub passed_checks: u32,
    pub pass_rate: f64,
    pub has_missing_content: bool,
    pub missing: MissingContent,
    pub success: bool,
}

/// Normalize text for comparison.
///
/// Removes extra whitespace and normalizes unicode characters,
/// LaTeX delimiters, and common symbol substitutions.
#[must_use]
pub fn normalize_text(text: &str) -> String {
    let mut result = text.trim().to_string();

    // Collapse whitespace
    if let Ok(re) = Regex::new(r"\s+") {
        result = re.replace_all(&result, " ").to_string();
    }
    // Normalize unicode spaces
    result = result
        .replace('\u{00A0}', " ")
        .replace('\u{2018}', "'")
        .replace('\u{2019}', "'")
        .replace('\u{201C}', "\"")
        .replace('\u{201D}', "\"")
        .replace('\u{00D7}', "x")
        .replace('\u{2192}', "->")
        .replace('\u{21A6}', "->")
        .replace('\u{2212}', "-");

    // Remove LaTeX delimiters
    result = result.replace("$$", "").replace('$', "");

    // Normalize LaTeX commands
    if let Ok(re) = Regex::new(r"\\times") {
        result = re.replace_all(&result, "x").to_string();
    }
    if let Ok(re) = Regex::new(r"\\to") {
        result = re.replace_all(&result, "->").to_string();
    }
    if let Ok(re) = Regex::new(r"\\displaystyle\s*") {
        result = re.replace_all(&result, "").to_string();
    }
    if let Ok(re) = Regex::new(r"\\text\{([^}]*)\}") {
        result = re.replace_all(&result, "$1").to_string();
    }
    result = result.replace("\\\\%", "%").replace("\\%", "%");
    result = result
        .replace("\\subseteq", "\u{2286}")
        .replace("\\in", "\u{2208}")
        .replace("\\emptyset", "\u{2205}");
    result = result.replace("^2", "\u{00B2}").replace("^n", "\u{207F}");

    // Handle \\mathbb{n}_0 case-insensitively
    if let Ok(re) = Regex::new(r"(?i)\\mathbb\{n\}_0") {
        result = re.replace_all(&result, "\u{2115}\u{2080}").to_string();
    }

    result.to_lowercase()
}

/// Normalize code for comparison (more lenient than text).
#[must_use]
pub fn normalize_code(text: &str) -> String {
    let mut result = text.trim().to_string();

    if let Ok(re) = Regex::new(r"\s+") {
        result = re.replace_all(&result, " ").to_string();
    }
    result = result
        .replace('\u{00A0}', " ")
        .replace('\u{00D7}', "x")
        .replace("$$", "")
        .replace('$', "");

    if let Ok(re) = Regex::new(r"\\times") {
        result = re.replace_all(&result, "x").to_string();
    }

    result.to_lowercase()
}

/// Verify that markdown contains the expected web page content.
#[must_use]
pub fn verify_markdown_content(
    web_content: &WebContent,
    markdown_text: &str,
    options: &VerifyOptions,
) -> VerifyResult {
    let normalized_markdown = normalize_text(markdown_text);
    let mut missing = MissingContent::default();
    let mut total_checks: u32 = 0;
    let mut passed_checks: u32 = 0;

    // Check title
    if let Some(ref title) = web_content.title {
        total_checks += 1;
        let normalized_title = normalize_text(title);
        if normalized_markdown.contains(&normalized_title) {
            passed_checks += 1;
        } else {
            missing.title = true;
        }
    }

    // Check headings
    for heading in &web_content.headings {
        total_checks += 1;
        let normalized = normalize_text(&heading.text);
        if normalized_markdown.contains(&normalized) {
            passed_checks += 1;
        } else {
            missing.headings.push(heading.text.clone());
        }
    }

    // Check paragraphs (sample first 5 and last 5)
    let paragraphs = &web_content.paragraphs;
    let first_five = paragraphs.iter().take(5);
    let last_five = if paragraphs.len() > 5 {
        paragraphs.iter().skip(paragraphs.len().saturating_sub(5))
    } else {
        paragraphs.iter().skip(paragraphs.len()) // empty
    };
    let paragraphs_to_check: Vec<&String> = first_five.chain(last_five).collect();

    for paragraph in &paragraphs_to_check {
        total_checks += 1;
        let normalized = normalize_text(paragraph);
        let words: Vec<&str> = normalized.split(' ').filter(|w| w.len() > 2).collect();
        let matching_words = words
            .iter()
            .filter(|word| normalized_markdown.contains(**word))
            .count();
        let match_rate = if words.is_empty() {
            0.0
        } else {
            matching_words as f64 / words.len() as f64
        };

        let substring_match = normalized.len() > 20
            && normalized_markdown.contains(&normalized[..normalized.len().min(50)]);

        if match_rate >= 0.6 || substring_match {
            passed_checks += 1;
        } else {
            let truncated = if paragraph.len() > 100 {
                format!("{}...", &paragraph[..100])
            } else {
                format!("{paragraph}...")
            };
            missing.paragraphs.push(truncated);
        }
    }

    // Check code blocks (fuzzy matching)
    let normalized_markdown_for_code = normalize_code(markdown_text);
    for code in &web_content.code_blocks {
        total_checks += 1;
        let normalized_code_full = normalize_code(code);

        let lines: Vec<&str> = code
            .lines()
            .map(|l| l.trim())
            .filter(|l| {
                l.len() > 3
                    && !Regex::new(r"^[{}\[\](),;]+$")
                        .map(|re| re.is_match(l))
                        .unwrap_or(false)
            })
            .collect();

        let matching_lines = lines
            .iter()
            .filter(|line| {
                let normalized_line = normalize_code(line);
                normalized_markdown_for_code.contains(&normalized_line)
            })
            .count();

        let match_rate = if lines.is_empty() {
            1.0
        } else {
            matching_lines as f64 / lines.len() as f64
        };

        if match_rate >= 0.6 || normalized_markdown_for_code.contains(&normalized_code_full) {
            passed_checks += 1;
        } else {
            let truncated = if code.len() > 100 {
                format!("{}...", &code[..100])
            } else {
                format!("{code}...")
            };
            missing.code_blocks.push(truncated);
        }
    }

    // Check list items (sample first 10)
    for item in web_content.list_items.iter().take(10) {
        total_checks += 1;
        let normalized = normalize_text(item);
        let words: Vec<&str> = normalized.split(' ').filter(|w| w.len() > 2).collect();
        let matching_words = words
            .iter()
            .filter(|word| normalized_markdown.contains(**word))
            .count();
        let match_rate = if words.is_empty() {
            0.0
        } else {
            matching_words as f64 / words.len() as f64
        };

        let substring_match = normalized.len() > 15
            && normalized_markdown.contains(&normalized[..normalized.len().min(40)]);

        if match_rate >= 0.6 || substring_match {
            passed_checks += 1;
        } else {
            let truncated = if item.len() > 100 {
                format!("{}...", &item[..100])
            } else {
                format!("{item}...")
            };
            missing.list_items.push(truncated);
        }
    }

    // Check blockquote formulas
    for formula in &web_content.blockquote_formulas {
        total_checks += 1;
        let normalized_formula = formula.split_whitespace().collect::<Vec<_>>().join(" ");

        // Extract key parts
        let cleaned = normalized_formula
            .replace("\\mathbf{", "")
            .replace("\\textbf{", "")
            .replace(['{', '}', '\\'], "");
        let key_parts: Vec<&str> = cleaned
            .split_whitespace()
            .filter(|part| part.len() > 1)
            .collect();

        // Find blockquote lines
        let blockquote_re = Regex::new(r"(?m)^>.*$").unwrap();
        let blockquote_lines: Vec<&str> = blockquote_re
            .find_iter(markdown_text)
            .map(|m| m.as_str())
            .collect();

        let mut found = false;
        for line in &blockquote_lines {
            if line.contains('$') {
                let matching_parts = key_parts
                    .iter()
                    .filter(|part| line.to_lowercase().contains(&part.to_lowercase()))
                    .count();
                if !key_parts.is_empty() && matching_parts >= key_parts.len().min(2) {
                    found = true;
                    break;
                }
                if line.contains(&normalized_formula)
                    || line.contains(formula.as_str())
                    || (formula.len() < 20 && line.contains(&formula.replace(' ', "")))
                {
                    found = true;
                    break;
                }
            }
        }

        if found {
            passed_checks += 1;
        } else {
            let truncated = if formula.len() > 100 {
                formula[..100].to_string()
            } else {
                formula.clone()
            };
            missing.blockquote_formulas.push(truncated);
        }
    }

    // Check for figure images
    if options.has_local_images {
        if let Some(expected) = options.expected_figures {
            total_checks += 1;
            let figure_re = Regex::new(
                r"(?i)!\[(?:\*\*)?(?:Figure|Рис\.?|Рисунок)\s*\d+[\s\S]*?\]\(images/figure-\d+\.(png|jpg)\)",
            )
            .unwrap();
            let figure_count = figure_re.find_iter(markdown_text).count() as u32;
            if figure_count >= expected {
                passed_checks += 1;
            } else {
                missing.images = expected - figure_count;
            }
        }
    }

    // Calculate results
    let pass_rate = if total_checks > 0 {
        f64::from(passed_checks) / f64::from(total_checks)
    } else {
        0.0
    };
    let has_missing_content = missing.title
        || missing.images > 0
        || !missing.headings.is_empty()
        || !missing.paragraphs.is_empty()
        || !missing.code_blocks.is_empty()
        || !missing.formulas.is_empty()
        || !missing.blockquote_formulas.is_empty()
        || !missing.list_items.is_empty();

    VerifyResult {
        total_checks,
        passed_checks,
        pass_rate,
        has_missing_content,
        success: !has_missing_content || pass_rate >= 0.85,
        missing,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_normalize_text_whitespace() {
        assert_eq!(normalize_text("  hello   world  "), "hello world");
    }

    #[test]
    fn test_normalize_text_unicode() {
        let result = normalize_text("test\u{00D7}value");
        assert!(result.contains('x'));
    }

    #[test]
    fn test_normalize_text_latex() {
        let result = normalize_text("$E = mc^2$");
        assert_eq!(result, "e = mc\u{00B2}");
    }

    #[test]
    fn test_normalize_code() {
        let result = normalize_code("  function()  {  }  ");
        assert_eq!(result, "function() { }");
    }

    #[test]
    fn test_verify_title_present() {
        let content = WebContent {
            title: Some("Hello World".to_string()),
            ..Default::default()
        };
        let result = verify_markdown_content(
            &content,
            "# Hello World\nSome content",
            &VerifyOptions::default(),
        );
        assert!(result.success);
        assert!(!result.missing.title);
    }

    #[test]
    fn test_verify_title_missing() {
        let content = WebContent {
            title: Some("Missing Title".to_string()),
            ..Default::default()
        };
        let result = verify_markdown_content(
            &content,
            "# Different Title\nSome content",
            &VerifyOptions::default(),
        );
        assert!(result.missing.title);
    }

    #[test]
    fn test_verify_headings() {
        let content = WebContent {
            headings: vec![
                Heading {
                    level: 2,
                    text: "Introduction".to_string(),
                },
                Heading {
                    level: 2,
                    text: "Conclusion".to_string(),
                },
            ],
            ..Default::default()
        };
        let result = verify_markdown_content(
            &content,
            "## Introduction\nText\n## Conclusion\nMore text",
            &VerifyOptions::default(),
        );
        assert_eq!(result.passed_checks, 2);
        assert!(result.missing.headings.is_empty());
    }

    #[test]
    fn test_verify_empty_content() {
        let content = WebContent::default();
        let result = verify_markdown_content(&content, "Some markdown", &VerifyOptions::default());
        assert!(result.success);
        assert_eq!(result.total_checks, 0);
    }
}
